### PR TITLE
Asya

### PR DESCRIPTION
**Pull request by asya-code**
AppointmentServices and Controller are done:
- list all appts
- view specific appt
- create new
- update existing
- delete existing

### CODE REVIEW

**Comment by GRISONRF** [comment]
At DoctorRepository it seems like you just fixed the typo on line 15 and imported its respective entity, correct?

I believe that's what we were talking about with Michael, that this type of fix should have its own PR, especially since the DoctorRepository folder is @ileanahi's responsibility.

So maybe let's talk about this at tonight's meeting before approving this PR?
@asya-code @williamsc51 @VeronicaO27

**Comment by ileanahi** [comment]
I agree that we should make separate branches that concern only one fix or update otherwise we'll have merge conflicts all over the place.

**Review by VeronicaO27 — APPROVED** [review]
Looks good to me

**Review by ileanahi — APPROVED** [review]
Looks good!

**Comment by williamsc51** [comment]
A few issues:

The constructor in the Appointments Entity that takes the AppointmentDto is not properly configured.
 https://github.com/ileanahi/doctors-office/blob/5cee516cca195555a3582ec556ecbfdaf3e6a425/src/main/java/com/capstone/doctorsOffice/entities/Appointment.java#L44
Current code:
` if (appointmentDto.getDay() != null){
            this.day = day;
 }`

You're assigning `day` to `this.day`
This is an error that prevent the `addNewAppointment` method from working

it should be.
` if (appointmentDto.getDay() != null){
            this.day = appointmentDto.getDay();
}`

I recommend updating the Appointment entity constructor to properly assign the values from the Dto.
And add it to this PR.

File: src/main/java/com/capstone/doctorsOffice/controllers/AppointmentController.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+    @GetMapping("/{appointmentId}")
+    public Optional<AppointmentDto> getAppointmentById(@PathVariable Long appointmentId) {
+        return appointmentService.getAppointmentById(appointmentId);
+    }
+
+    @PostMapping("/new-appointment")
+    public List<String> addNewAppointment(@RequestBody AppointmentDto appointmentDto) {
+        return appointmentService.addNewAppointment(appointmentDto);
+    }
+
```
Comment (williamsc51) [inline src/main/java/com/capstone/doctorsOffice/controllers/AppointmentController.java:34]: A few issues:

The constructor in the Appointments Entity that takes the AppointmentDto is not properly configured.

[doctors-office/src/main/java/com/capstone/doctorsOffice/entities/Appointment.java](https://github.com/ileanahi/doctors-office/blob/5cee516cca195555a3582ec556ecbfdaf3e6a425/src/main/java/com/capstone/doctorsOffice/entities/Appointment.java#L44)
Line 44 in [5cee516](https://github.com/ileanahi/doctors-office/commit/5cee516cca195555a3582ec556ecbfdaf3e6a425)
 public Appointment(AppointmentDto appointmentDto){ 

Current code:
 if (appointmentDto.getDay() != null){ this.day = day; }
You're assigning day to this.day
This is an error that prevent the addNewAppointment method from working

it should be.
 if (appointmentDto.getDay() != null){ this.day = appointmentDto.getDay(); }

I recommend updating the Appointment entity constructor to properly assign the values from the Dto.
And add it to this PR.

File: src/main/java/com/capstone/doctorsOffice/controllers/AppointmentController.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+    @DeleteMapping("/{appointmentId}")
+    public void deleteAppointment(@PathVariable Long appointmentId){
+        appointmentService.deleteAppointment(appointmentId);
+    }
+
+    @PutMapping("/{appointmentId}")
+    public void updateAppointment(@PathVariable Long appointmentId,
+                                  @RequestBody AppointmentDto appointmentDto){
+        appointmentService.updateAppointment(appointmentDto);
+    }
```
Comment (williamsc51) [inline src/main/java/com/capstone/doctorsOffice/controllers/AppointmentController.java:44]: You need to pass the path variable `appointmentId` to the service to look up the appointment,
should be:
`appointmentService.updateAppointment(appointmentId, appointmentDto);`

File: src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+    @Override
+    @Transactional
+    public void updateAppointment(AppointmentDto appointmentDto){
+        Optional<Appointment> appointmentOptional = appointmentRepository.findById(appointmentDto.getAppointmentId());
+        appointmentOptional.ifPresent(appointment -> {
+            appointment.setTime(appointmentDto.getTime());
+            appointment.setDay(appointmentDto.getDay());
+        });
+    }
+
```
Comment (williamsc51) [inline src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java:50]: The updateAppointment method should take two arguments
`public void updateAppointment(Long  appointment, AppointmentDto appointmentDto)`
then use the `appointmentId` to look up the appointment 
`Optional<Appointment> appointmentOptional = appointmentRepository.findById(appointmentId);`

**Review by williamsc51 — CHANGES_REQUESTED** [review]

File: src/main/java/com/capstone/doctorsOffice/services/AppointmentService.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+public interface AppointmentService {
+    @Transactional
+    List<String> addNewAppointment(AppointmentDto appointmentDto);
+
+    @Transactional
+    void deleteAppointment(Long appointmentId);
+
+    @Transactional
+    void updateAppointment(AppointmentDto appointmentDto);
+
```
Comment (williamsc51) [inline src/main/java/com/capstone/doctorsOffice/services/AppointmentService.java:18]: update function signature in interface to take the additional parameter
void updateAppointment(Long  appointment, AppointmentDto appointmentDto)

**Review by williamsc51 — CHANGES_REQUESTED** [review]

File: src/main/java/com/capstone/doctorsOffice/services/AppointmentService.java (hunk 4 above)
Comment (asya-code) [reply src/main/java/com/capstone/doctorsOffice/services/AppointmentService.java:18]: Done

File: src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    public void deleteAppointment(Long appointmentId){
+        Optional<Appointment> appointmentOptional = appointmentRepository.findById(appointmentId);
+        appointmentOptional.ifPresent(appointment -> appointmentRepository.delete(appointment));
+    }
+
+    @Override
+    @Transactional
+    public void updateAppointment(Long appointmentId, AppointmentDto appointmentDto){
+        Optional<Appointment> appointmentOptional = appointmentRepository.findById(appointmentDto.getAppointmentId());
+        appointmentOptional.ifPresent(appointment -> {
```
Comment (williamsc51) [inline src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java:45]: We need to use the `appointmentId` to look up the entity. The Dto may not have the ID.
should be:
`Optional<Appointment> appointmentOptional = appointmentRepository.findById(appointmentId);`

**Review by williamsc51 — CHANGES_REQUESTED** [review]

File: src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java (hunk 5 above)
Comment (asya-code) [reply src/main/java/com/capstone/doctorsOffice/services/AppointmentServiceImpl.java:45]: done

**Review by williamsc51 — APPROVED** [review]

**Pull request merged**